[Pipe] Add support client reloading

faustinoaq opened the pull request:
### Description of the Change

Hi team 😄 Happy Christmas! 🎉 

- Add Hot reloading for CSS
- Add reloading for files on public folder
- Allow reloading browser after server reload

![ezgif-5-1c180670af](https://user-images.githubusercontent.com/3067335/34326676-126266d6-e881-11e7-89d9-015868d02729.gif)

_`Backend reloading, client reloading and hot reloading (preserving state)`_

### Alternate Designs

Using `Amber::Tasks` ¿? @eliasjpr 

### Benefits

Fix  #347 definitely

### Possible Drawbacks

- ~~`Amber::Pipe::Reload` keeps enabled on production environments~~
- `Amber::Pipe::Reload` modifies `"content-type"`, so json/text/other responses will have injected code too, a posible solution would be adding `Pipe::Reload` after Controllers processing, then injected code only exists if `"content-type"` is equal to `"text/html"` instead of enforcing `"text/html"`

robacarp commented:
This is great!

eliasjpr commented on src/amber/support/client_reload.cr:
```diff
@@ -0,0 +1,94 @@
+      watch(["public/**/*"]) do |event|
+        event.on_change do |files|
+          files.each do |file, timestamp|
+            puts "🔃  watching file: ./#{file}"
```
This should use the new logger `Amber.logger.puts "watching file: ./#{file}", "Watcher`

faustinoaq replied:
Yeah, I thought you weren't using logger on watcher, I just realize latest amber `v0.5.0` was republished to fix logs. 😅 

Let me fix this 👍

faustinoaq replied:
Done!

eliasjpr commented on src/amber/support/client_reload.cr:
```diff
@@ -0,0 +1,94 @@
+    end
+
+    private def watch_files
+      puts "🔃  Your ClientBot is vigilant. beep-boop..."
```
I would remove this log is not useful info

faustinoaq replied:
Done!

eliasjpr commented on shard.yml:
```diff
@@ -68,3 +68,7 @@ dependencies:
+
+  watcher:
+    github: faustinoaq/watcher
+    version: ~> 0.2.0
```
New line missing

faustinoaq replied:
Done!

eliasjpr commented on src/amber/router/pipe/reload.cr:
```diff
@@ -0,0 +1,26 @@
+      end
+
+      def call(context : HTTP::Server::Context)
+        context.response.print Support::ClientReload::INJECTED_CODE
```
Does print close the response?

faustinoaq replied:
@eliasjpr I don't think so

What other method should I use?

eliasjpr replied:
If it does not close the response thats good ccause `Pipe::Last` has to run.

drujensen commented on shard.yml:
```diff
@@ -68,3 +68,7 @@ dependencies:
     version: ~> 0.8.3
+
+  watcher:
+    github: faustinoaq/watcher
```
Is this a replacement for sentry? If so, can we remove sentry?

faustinoaq replied:
I removed both 👍 less dependencies, less to worry about

drujensen commented on src/amber/support/client_reload.cr:
```diff
@@ -0,0 +1,94 @@
+            var head = document.getElementsByTagName('head')[0];
+            for (var i = 0; i < sheets.length; ++i) {
+              var elem = sheets[i];
+              head.removeChild(elem);
```
Should we only remove and add the link if its a `stylesheet` or should we do this for all links?

faustinoaq replied:
I just copied a working code from nodejs package `live-server`, What you suggest we could use? 😅

drujensen replied:
The logic here gets all the `links`, removes it from the head, determines if its a `stylesheet` on line 62.  If it is, it adds a `_cacheOverride={Date()}` to the link to break the cache.  Then it adds the link back into the head.

What I'm wondering if we need to remove it then append it back on #67 if its not a stylesheet.  Why remove it and add it back if your not going to add a `cacheOverride`?

What if we only remove and add it back if its a stylesheet?  Something like this:
```
if (elem.href && typeof rel != 'string' || rel.length == 0 || rel.toLowerCase() == 'stylesheet') {
  head.removeChild(elem);
  var url = elem.href.replace(/(&|\\?)_cacheOverride=\\d+/, '');
  elem.href = url + (url.indexOf('?') >= 0 ? '&' : '?') + '_cacheOverride=' + (new Date().valueOf());
  head.appendChild(elem);
}
```

faustinoaq replied:
Done!

drujensen commented on src/amber/cli/templates/app/shard.yml.ecr:
```diff
@@ -57,3 +57,7 @@ development_dependencies:
   garnet_spec:
     github: amberframework/garnet-spec
     version: ~> 0.1.1
+
```
If Amber requires this as a dependency, do we need it here as well?

faustinoaq replied:
You're right, let me fix this 👍

faustinoaq commented:
Hi team, I have some comments about amber and this PR:

> Does this replaces the npm dependency?

@eliasjpr A backend asset proccess like NPM is still needed to compile and bundle things like TypeScript and SCSS files. This code reloading feature just check file changes on public files. You can use Vanilla JS and CSS without NPM.

> Do we need to remove the npm run script from in process runner? See

I think NPM inside `amber watch` is nice, but what about other tools like yarn, or what happens if a developers want to use webpack-dev-server with proxy or rollupjs instead.

We can just disable `Amber::Pipe::Reload` by commenting this on pipeline, but How can we disable NPM on `amber watch`?

Maybe we can use flags like `--npm` and `--yarn` to support them optionally.

@eliasjpr Before you include NPM inside `amber watch` I have been using an extra terminal for assets stuff. Maybe backend developers like all in one place working by default, but I think front-end developers have so many JS tools, so enforcing them to use NPM only isn't good 😅

Other comments:

1) Sentry is not needed anymore, It has been embedded completely inside amber, see `Sentry::ProcessRunner` implemented on #458 
2) Amber `v0.5.0` is running NPM install on every reload, this is a unwanted behavior, so I added a flag to run it at start only.
3) I removed NPM warnings using `--loglevel=error` option

marksiemers commented on src/amber/support/client_reload.cr:
```diff
@@ -29,19 +34,38 @@ module Amber::Support
+        elsif FILE_TIMESTAMPS[file]?.nil?
+          FILE_TIMESTAMPS[file] = timestamp
+          file_changed = true
+          Amber.logger.puts "Watching file: ./#{file.colorize(:light_gray)}", "Watcher", :light_gray
```
Do the messages need to be different - one saying `Watching file: ...` the other, looks the same other than that prefix.

Could this be simplified?
```crystal
private def scan_files
  Dir.glob(["public/**/*"]) do |file|
    timestamp = get_timestamp(file)
    msg_prefix = FILE_TIMESTAMPS[file]? ? "" : "Watching file: "

    if FILE_TIMESTAMPS[file]? != timestamp
      FILE_TIMESTAMPS[file] = timestamp
      Amber.logger.puts "#{msg_prefix}./#{file.colorize(:light_gray)}", "Watcher", :light_gray
      check_file(file)
    end
  end
end
```

faustinoaq replied:
Thank you! done 👍

faustinoaq commented on src/amber/router/pipe/reload.cr:
```diff
@@ -0,0 +1,27 @@
+      end
+
+      def call(context : HTTP::Server::Context)
+        if context.format == "html"
```
BTW, @eliasjpr I found another issue here 😅 context.format always return "html", so the following code, gets injected too:

```crystal
class HomeController < ApplicationController
  def index
    context.response.headers["context-header"] = "application/json"
    {foo: "bar"}.to_json
  end
end
```

How can I fix this?

eliasjpr replied:
The  `Accept` request header must be set in order for it to return the correct format otherwise it will return `html` by default.

eliasjpr replied:
Currently working on a fix for this

faustinoaq commented on src/amber/cli/helpers/process_runner.cr:
```diff
@@ -26,10 +27,16 @@ module Sentry
+      elsif !@app_running
+        log "Compile time errors detected. Shutting down..."
+        exit 1
+      end
```
Done!  👍

drujensen commented:
> Maybe we can use flags like --npm and --yarn to support them optionally.

I think we should consider changing `amber watch` to be configurable similar to `foreman`using a `Procfile`. 

We specify a list of processes providing what files to watch and what command to run if any of the files change.  This will allow `amber watch` to be used for any toolset the user wants and it prevents the dependency on `npm`.  For example, they may want to watch the `/db` directory and restart `crystal db migrate`.  Or they will want to watch `/jobs` directory and restart `crystal run src/sidekiq.cr`.  These are just a couple examples.

faustinoaq commented:
I'm having some issues on TravisCI:

Passing 👉 https://travis-ci.org/amberframework/amber/jobs/321150453#L1606 ✔️ 
Failing 👉 https://travis-ci.org/amberframework/amber/jobs/321967054#L1620 ❎ (same code)

```crystal
require "../../../../spec_helper"

module Amber
  module Pipe
    describe Reload do
      it "client should contain injected code" do
        reload = Reload.new
        pipeline = Pipeline.new
        request = HTTP::Request.new("GET", "/reload")

        Amber::Server.router.draw :web do
          get "/reload", HelloController, :index
        end

        pipeline.build :web do
          plug Amber::Pipe::Reload.new
        end

        reload.next = ->(context : HTTP::Server::Context) { "Hello World!" }
        response = create_request_and_return_io(reload, request)

        response.body.should contain "Code injected by Amber Framework"
      end
    end
  end
end
```

faustinoaq commented:
@drujensen Maybe we can use `.amber.yml` adding some extra information, something like:

```yml
type: app
database: sqlite
language: slang
model: granite
watch:
  server:
    files: ["src/**/*.cr", "src/**/*.ecr"]
    command: shards build
  client:
    files: "public/**/*"
    command: npm run watcher
  db:
    files: "db/**/*"
    command: amber db migrate
  jobs:
    files: "src/jobs/*"
    command: crystal run src/sidekiq.cr
```

WDYT?

robacarp commented:
A dedicated Procfile has the added benefit of being an accepted standard on several existing servers and hosts

drujensen commented:
@faustinoaq yes, that is inline with what I was thinking.

@robacarp I wish the format for `Procfile` was more flexible.  Unfortunately, without the ability to specify which files to watch, we wouldn't be able to use it.

faustinoaq commented:
I think we can implement an `amber watch` file configuration in other PR 😅 

Something like:

```yaml
watch:
  server:
    files: ["src/**/*.cr", "src/**/*.ecr"]
    command: shards build
  client:
    files: "public/**/*"
    command: ["yarn install", "yarn run watch"]
  db:
    files: "db/**/*"
    command: amber db migrate
  jobs:
    files: "src/jobs/*"
    command: crystal run src/sidekiq.cr
```

WDYT?

BTW, I'm trying to pass Travis, but I don't know why how to fix/implement [this](https://github.com/amberframework/amber/pull/457/files#diff-946aa78ba53835a4d30a60e6737989a6R6) spec

eliasjpr commented:
@faustinoaq I think this commit https://github.com/amberframework/amber/pull/457/commits/a3826434900d6fc3e07bfb31d143ad0547a31aa6 fixes the specs. Let me know what you think.

eliasjpr commented:
Also I think you might want to rebase interactive to drop/squash some commits before merging

elorest commented on src/amber/cli/templates/app/config/routes.cr:
```diff
@@ -7,6 +7,11 @@ Amber::Server.configure do |app|
     plug Amber::Pipe::CSRF.new
+
+    # Reload clients browsers (development only)
+    if Amber.env.development?
```
Does this logic work? Seems to be a conditional around a macro that generates code.

faustinoaq replied:
I tried it and work as expected

faustinoaq commented:
> @faustinoaq I think this commit a382643 fixes the specs. Let me know what you think.

Yeah, Thank you! 🎉 

> Also I think you might want to rebase interactive to drop/squash some commits before merging

I agree 👍

drujensen commented:
I added #476 to make `amber watch` configurable

faustinoaq commented:
> Looks good. Don't think it's quite ready yet though is it?

@elorest Yeah, Still wondering How to fix [this](https://github.com/amberframework/amber/pull/457#discussion_r158740400) 😅 

Currently this implementation always injects the client code on every response because `context.format` always return `html` inside `Amber::Pipe::Reload`.

Finally, Another PR is needed to achieve #476
